001-resources-v2/proposal.md

```diff
@@ -0,0 +1,626 @@
+
+  Required fields for this event:
+
+  * `space`: The space the version is in.
+  * `version`: The version object.
+  * `metadata`: A list of JSON objects with `name` and `value`, shown to the
+    user. Note that this is return by both `put` and `check`, because there's a
+    chance that `put` produces a version that wouldn't normally be discovered
+    by `check`.
+
+* `deleted`: Emitted when a version has been deleted. The version record will
+  remain in the database for archival purposes, but it will no longer be a
+  candidate for any builds.
+
+  Required fields for this event:
+
+  * `space`: The space the version is in.
+  * `version`: The version object.
+
+Because the space is included on each event, `put` allows new spaces to be
+generated dynamically based on params and/or the bits in its working directory
+and propagated to the rest of the pipeline.
+
+
+# Examples
+
+## Resource Implementations
+
+I've started cooking up new resources using this interface. I've left `TODO`s
+for parts that need more thinking or discussion. Please leave comments!
+
+### `git`
+
+[Code](https://github.com/vito/rfcs/tree/resources-v2/01-resources-v2/git-example)
+
+This resource models the original `git` resource. It represents each branch as a space.
+
+### `semver-git`
+
+[Code](https://github.com/vito/rfcs/tree/resources-v2/01-resources-v2/semver-example)
+
+This is a whole new semver resource intended to replace the original `semver`
+resource with a better model that supports concurrent version lines (i.e.
+supporting multiple major/minor releases with patches). It does this by managing
+tags in an existing Git repository.
+
+### `s3`
+
+[Code](https://github.com/vito/rfcs/tree/resources-v2/01-resources-v2/s3-example)
+
+This resource models the original `s3` resource. Only regex versions were
+implemented, each space corresponds to a major.minor version. For example, 1.2.0
+and 1.2.1 is the same space but 1.3.0 is a different space. Single numbers are
+also supported with default minor of 0. The default space is set to the latest
+minor version.
+
+
+## Pipeline Usage
+
+TODO:
+
+- Pull Requests
+- Feature branches
+- Build matrixes
+- Generating branches (and propagating them downstream)
+- Semver artifacts
+- Fanning out against multiple IaaSes
+- Pool resource?
+- BOSH deploys
+
+
+# Summary of Changes
+
+## Overarching Changes
+
+* Add an `info` script which returns a JSON object indicating the supported
+  interfaces, their protocol versions, and any other interface-specific
+  meta-configuration (for example, which commands to execute for the
+  interface's hooks).
+
+* The first supported interface will be called `artifacts`, and its version
+  will start at `2.0` as it's really the next iteration of the existing
+  "resources" concept, but with a more specific name.
+
+* There are no more hardcoded paths (`/opt/resource/X`) - instead there's the
+  single `info` entrypoint, which is run in the container's working directory.
```

Comment:
`info` semantically implies read-only. If it is going to replace also the `out`/`put`, maybe a more adequate name could be `command` or similar.

Reply:
Confusing wording - it's read-only, I meant `entrypoint` from a discovery standpoint (invoke `info` to learn the later commands to run), not an execution standpoint. Will amend.